File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.swift
import Foundation

extension Int {
    func isPrime() -> Bool {
        if self < 2 { return false }
        var i = 2
        while i * i <= self {
            if self % i == 0 { return false }
            i += 1
        }
        return true
    }
}

func permutation(_ arr: [String], _ k: Int) -> [String] {
    
    var answer: [String] = []
    var selected: [String] = []
    var isUsed = [Bool](repeating: false, count: arr.count)
    
    func select(_ s: Int) {
        
        if s == k {
            answer.append(selected.reduce("", +))
        }
        
        for i in 0..<arr.count {
            if !isUsed[i] {
                selected.append(arr[i])
                isUsed[i] = true
                select(s+1)
                isUsed[i] = false
                selected.removeLast()
            }
        }
    }
    select(0)
    
    return answer
}

func solution(_ numbers:String) -> Int {
    
    var nums = Array(numbers).map { String($0) }
    
    var primeSet = Set<Int>()
    
    for i in 1...nums.count {
        let num = permutation(nums, i).map { Int($0)! }
        
        for n in num {
            if n.isPrime() {
                primeSet.insert(n)
            }
        }
    }
    
    return primeSet.count
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
# 한자리 숫자

# 7! -> 4320개 경우의 수

from itertools import permutations

def isPrime(num):
    
    if num < 2:
        return False
    
    i = 2
    while i * i <= num:
        if num % i == 0:
            return False
        i += 1
    return True

def solution(numbers):
    
    numbers = list(map(int, numbers))
    
    s = set()
    
    for r in range(1, len(numbers) + 1):
        for p in permutations(numbers, r):
            num = int(''.join(map(str, p)))
              
            if isPrime(num):
                s.add(num)

    return len(s)